boards/nordic/nrf52840dk/src/main.rs

```diff
@@ -305,31 +300,86 @@ impl KernelResources<nrf52840::chip::NRF52<'static, Nrf52840DefaultPeripherals<'
+    // INITIAL SETUP
+    //--------------------------------------------------------------------------
+
+    // Apply errata fixes and enable interrupts.
     nrf52840::init();
 
+    // Set up peripheral drivers. Called in separate function to reduce stack
+    // usage.
     let nrf52840_peripherals = create_peripherals();
 
-    // set up circular peripheral dependencies
+    // Set up circular peripheral dependencies.
     nrf52840_peripherals.init();
     let base_peripherals = &nrf52840_peripherals.nrf52;
 
+    // Choose the channel for serial output. This board can be configured to use
+    // either the Segger RTT channel or via UART with traditional TX/RX GPIO
+    // pins.
     let uart_channel = if USB_DEBUGGING {
-        // Initialize early so any panic beyond this point can use the RTT memory object.
+        // Initialize early so any panic beyond this point can use the RTT
+        // memory object.
         let mut rtt_memory_refs = components::segger_rtt::SeggerRttMemoryComponent::new()
             .finalize(components::segger_rtt_memory_component_static!());
 
-        // XXX: This is inherently unsafe as it aliases the mutable reference to rtt_memory. This
-        // aliases reference is only used inside a panic handler, which should be OK, but maybe we
-        // should use a const reference to rtt_memory and leverage interior mutability instead.
+        // XXX: This is inherently unsafe as it aliases the mutable reference to
+        // rtt_memory. This aliases reference is only used inside a panic
+        // handler, which should be OK, but maybe we should use a const
+        // reference to rtt_memory and leverage interior mutability instead.
         self::io::set_rtt_memory(&mut *rtt_memory_refs.get_rtt_memory_ptr());
 
         UartChannel::Rtt(rtt_memory_refs)
     } else {
         UartChannel::Pins(UartPins::new(UART_RTS, UART_TXD, UART_CTS, UART_RXD))
     };
 
+    // Setup space to store the core kernel data structure.
     let board_kernel = static_init!(kernel::Kernel, kernel::Kernel::new(&PROCESSES));
 
+    // Create (and save for panic debugging) a chip object to setup low-level
+    // resources (e.g. MPU, systick).
+    let chip = static_init!(
+        nrf52840::chip::NRF52<Nrf52840DefaultPeripherals>,
+        nrf52840::chip::NRF52::new(nrf52840_peripherals)
+    );
+    CHIP = Some(chip);
+
+    // Do nRF configuration and setup. This is shared code with other nRF-based
+    // platforms.
+    nrf52_components::startup::NrfStartupComponent::new(
+        false,
+        BUTTON_RST_PIN,
+        nrf52840::uicr::Regulator0Output::DEFAULT,
+        &base_peripherals.nvmc,
+    )
+    .finalize(());
+
+    // Configure kernel debug GPIOs as early as possible.
+    kernel::debug::assign_gpios(
+        Some(&nrf52840_peripherals.gpio_port[LED1_PIN]),
+        Some(&nrf52840_peripherals.gpio_port[LED2_PIN]),
+        Some(&nrf52840_peripherals.gpio_port[LED3_PIN]),
+    );
```

Comment:
Can this move to ~line 317 (right after the `nrf52840_peripherals` object is created? The comment is a bit confusing as-is — why is this "as early as possible"; the NrfStartupComponent doesn't touch the gpios, just the `.nvmc`?)

Reply:
Can do!